File: leetcode/c++/0125-valid-palindrome.cc
// https://leetcode.com/problems/valid-palindrome/
// Easy

#include<string>

using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int lo = -1, len = s.length(), hi = len;
        bool loAl = false, hiAl = false;
        while (lo < hi) {
            while (lo < len && !isalnum(s[++lo])) { }
            if (lo == len) return true;
            while (!isalnum(s[--hi])) { }

            if (tolower(s[lo]) != tolower(s[hi])) {
                return false;
            }
        }
        return true;
    }
};

File: leetcode/python/0125-valid-palindrome.py
# https://leetcode.com/problems/valid-palindrome/
# Easy

# Short solution
class Solution:
    def isPalindrome(self, s: str) -> bool:
        chars = "".join([c for c in s if c.isalnum()]).lower()
        return chars == chars[::-1]
# Using filter
class Solution:
    def isPalindrome(self, s: str) -> bool:
        chars = "".join(filter(str.isalnum, s.lower()))
        return chars == chars[::-1]

# Two pointer solution
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        lo, hi = 0, len(s)-1
        while lo < hi:
            if not s[lo].isalnum():
                lo += 1
            elif not s[hi].isalnum():
                hi -= 1
            elif s[lo] != s[hi]:
                return False
            else:
                lo += 1
                hi -= 1
        return True